trainers/stage_2_trainer.py

```diff
@@ -0,0 +1,168 @@
+import numpy as np
+from tqdm import trange
+import tensorflow as tf
+from trainers.base_trainer import Trainer
+from utils.data_helpers import tensors_from_sample
+
+
+class Stage1Trainer(Trainer):
+    """ Trainer which feeds in text as input to the GAN to generate images """
+    def __init__(self, model, batch_size, save_location,
+                 save_every, save_best_after, show_progress_bar=True, **kwargs):
+        """ Initialise a model trainer for iamge data.
+            Arguments:
+            model: models.ConditionalGAN
+                The model to train
+            batch_size: int
+                The number of samples per mini-batch
+            save_location: str
+                The directory in which to save all
+                results from training the model.
+        """
+        super().__init__(model, batch_size, save_location, save_every, save_best_after, show_progress_bar)
+        self.num_embeddings = kwargs.get('num_embeddings')
+        self.num_samples = kwargs.get('num_samples')
+        self.noise_size = kwargs.get('noise_size')
+        self.augment = kwargs.get('augment')
+        self.stage_1_generator = kwargs.get('stage_1_generator')
+
+    def train_epoch(self, train_loader, epoch_num):
+        """ Training operations for a single epoch """
+        acc_generator_loss = 0
+        acc_discriminator_loss = 0
+        text_embedding_size = train_loader.dataset_object.text_embedding_dim
+        kwargs = dict(
+            desc="Epoch {}".format(epoch_num + 1),
+            leave=False,
+            disable=not self.show_progress_bar
+        )
+        with trange(len(train_loader), **kwargs) as t:
+            for batch_idx, sample in enumerate(train_loader.parsed_subset):
+                batch_size = len(sample['text'].numpy())
+                _, _, image_large, wrong_image_large, text_tensor = tensors_from_sample(
+                    sample, batch_size, text_embedding_size, self.num_samples, self.augment
+                )
+
+                with tf.GradientTape() as generator_tape, tf.GradientTape() as discriminator_tape:
+                    # Forward pass the stage 1 generator to obtain small fake images
+                    noise_z = tf.random.normal((batch_size, self.noise_size))
+                    fake_images_small, _, _ = self.stage_1_generator([text_tensor, noise_z], training=False)
+
+                    fake_images_large, mean, log_sigma = self.model.generator(
+                        [text_tensor, noise_z, fake_images_small], training=True
```

Comment:
yeah just checked, no noise. Only need to pass in [fake_images_small, text_tensor] in that order.

Reply:
Perfect, thanks. Their figure 2 is quite misleading then

Reply:
![stage2](https://user-images.githubusercontent.com/13699844/81475484-cc1df680-920c-11ea-9fcf-5f8ad1d7bf08.png)
the only things I see going into stage 2 are the embedding and generated images unless I am missing something?

Reply:
Inside the conditioning augmentation block there is a noise generation mechanism if we use the same definition as given in stage 1

Reply:
oh ok yeah, but that is not noise_z though